File: 803-cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<int> prices;
        for (int i = 0; i < n; i++) prices.push_back(INT_MAX);

        prices[src] = 0;

        for (int i = 0; i < k+1; i++){
            vector<int> tempPrices;
            tempPrices.assign(prices.begin(), prices.end());
            for (vector<int> flight: flights){
                if (prices[flight[0]] == INT_MAX) continue;
                if (prices[flight[0]] + flight[2] < tempPrices[flight[1]]) tempPrices[flight[1]] = prices[flight[0]] + flight[2];
            }
            prices = tempPrices;
        }

        return prices[dst] == INT_MAX ? -1 :  prices[dst];
    }
};

File: 803-cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        price = [float("inf")] * n
        price[src] = 0
        for _ in range(k + 1):
            tempPrice = price.copy()
            for s, d, p in flights:
                if price[s] == float("inf"): continue
                if price[s] + p < tempPrice[d]:
                    tempPrice[d] = price[s] + p
            price = tempPrice.copy()
        
        return -1 if price[dst] == float("inf") else price[dst]